smart_open/smart_open_lib.py

```diff
@@ -233,6 +232,60 @@ def open(
     return decoded
 
 
+def _get_binary_mode(mode_str):
+    #
+    # https://docs.python.org/3/library/functions.html#open
+    #
+    # The order of characters in the mode parameter appears to be unspecified.
+    # The implementation follows the examples, just to be safe.
+    #
+    mode = list(mode_str)
+    binmode = []
+
+    if 't' in mode and 'b' in mode:
+        raise ValueError("can't have text and binary mode at once")
+
+    counts = [mode.count(x) for x in 'rwa']
+    if sum(counts) > 1:
+        raise ValueError("must have exactly one of create/read/write/append mode")
+
+    def transfer(char):
+        binmode.append(mode.pop(mode.index(char)))
+
+    if 'a' in mode:
+        transfer('a')
+    elif 'w' in mode:
+        transfer('w')
+    elif 'r' in mode:
+        transfer('r')
+    else:
+        raise ValueError(
+            "Must have exactly one of create/read/write/append "
+            "mode and at most one plus"
```

Comment:
This doesn't match built-in `open`: `open(fname)` works, with a default `mode='r'`.

Reply:
By this time, we've already applied the default, so that is covered.

The error handling here is to cover `open(filename, '+')`, which will trigger that error with the built-in open.